src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/BlockMerkleRootRule.cs

```diff
@@ -70,63 +68,58 @@ public static uint256 BlockMerkleRoot(Block block, out bool mutated)
         public static uint256 ComputeMerkleRoot(List<uint256> leaves, out bool mutated)
         {
-            var branch = new List<uint256>();
-
             mutated = false;
-            if (leaves.Count == 0)
-                return uint256.Zero;
+            if (leaves.Count == 0) return uint256.Zero;
 
-            // count is the number of leaves processed so far.
-            uint count = 0;
+            var branch = new List<uint256>();
 
-            // inner is an array of eagerly computed subtree hashes, indexed by tree
+            // subTreeHashes is an array of eagerly computed subtree hashes, indexed by tree
             // level (0 being the leaves).
-            // For example, when count is 25 (11001 in binary), inner[4] is the hash of
-            // the first 16 leaves, inner[3] of the next 8 leaves, and inner[0] equal to
-            // the last leaf. The other inner entries are undefined.
-            var inner = new uint256[32];
+            // For example, when count is 25 (11001 in binary), subTreeHashes[4] is the hash of
+            // the first 16 leaves, subTreeHashes[3] of the next 8 leaves, and subTreeHashes[0] equal to
+            // the last leaf. The other subTreeHashes entries are undefined.
+            var subTreeHashes = new uint256[32];
 
-            for (int i = 0; i < inner.Length; i++)
-                inner[i] = uint256.Zero;
+            for (int i = 0; i < subTreeHashes.Length; i++)
+                subTreeHashes[i] = uint256.Zero;
 
             // Which position in inner is a hash that depends on the matching leaf.
             int matchLevel = -1;
+            uint processedLeavesCount = 0;
 
-            // First process all leaves into 'inner' values.
-            while (count < leaves.Count)
+            // First process all leaves into subTreeHashes values.
+            while (processedLeavesCount < leaves.Count)
             {
-                uint256 h = leaves[(int)count];
+                uint256 currentLeaveHash = leaves[(int)processedLeavesCount];
                 bool match = false;
-                count++;
+                processedLeavesCount++;
                 int level;
 
-                // For each of the lower bits in count that are 0, do 1 step. Each
-                // corresponds to an inner value that existed before processing the
+                // For each of the lower bits in processedLeavesCount that are 0, do 1 step. Each
+                // corresponds to an subTreeHash value that existed before processing the
                 // current leaf, and each needs a hash to combine it.
-                for (level = 0; (count & (((uint)1) << level)) == 0; level++)
+                for ( level = 0; (processedLeavesCount & (((uint)1) << level)) == 0; level++)
+                //for (level = 0; (processedLeavesCount % ((2^level) -1)) == 0; level++)
```

Comment:
Please remove commented out code

Reply:
I left it here on purpose because I thought it was really not obvious what this loop was doing...

Reply:
Oh, ok. 

So probably it will be a good idea to replace the commented out code with a text comment so it's obvious that this line is an explanation, not an alternative implementation.

Reply:
if something is not clear, make a normal comment instead

Reply:
but there is comment already and it seems that the rewrite of `(processedLeavesCount % ((2^level) -1)) == 0` does not make much sense, we are doing an action for every unset bit inside of `count` up until the first set bit, while modulo operation is logically something else

this is more like
count is `101101000`, then you execute the loop 3 times (as count ands with 3 zero bits)

Reply:
haha so I definitely don't get that way of writing "%" but it feels like it is equivalent though. I just find it easier to imagine the loop when I read my comment than when I see operations on bits

Reply:
just try to think what happens in your for:
```
for (level = 0; (processedLeavesCount % ((2^level) -1)) == 0; level++)
```

so imagine level is 0, then you have `(pLC % ((2 ^ 0) -1 ))` which is `pLC % (1 - 1)`, which means you are dividing by zero

Reply:
it is not clearer, it is wrong, pls remove it